svf/include/Util/SVFBugReport.h

```diff
@@ -0,0 +1,324 @@
+    {
+        return bug->getBugType() == GenericBug::PARTIALBUFOVERFLOW;
+    }
+};
+
+class NeverFreeBug : public GenericBug{
+public:
+    NeverFreeBug(const EventStack &bugEventStack):
+          GenericBug(GenericBug::NEVERFREE, bugEventStack){  };
+
+    cJSON *getBugDescription();
+    void printBugToTerminal();
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::NEVERFREE;
+    }
+};
+
+class PartialLeakBug : public GenericBug{
+protected:
+    Set<std::string> conditionalFreePath;
+
+public:
+    PartialLeakBug(const EventStack &bugEventStack, Set<std::string> conditionalFreePath):
+          GenericBug(GenericBug::PARTIALLEAK, bugEventStack), conditionalFreePath(conditionalFreePath){ }
+
+    cJSON *getBugDescription();
+    void printBugToTerminal();
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::PARTIALLEAK;
+    }
+};
+
+class DoubleFreeBug : public GenericBug{
+protected:
+    Set<std::string> doubleFreePath;
+
+public:
+    DoubleFreeBug(const EventStack &bugEventStack, Set<std::string> doubleFreePath):
+          GenericBug(GenericBug::PARTIALLEAK, bugEventStack), doubleFreePath(doubleFreePath){ }
+
+    cJSON *getBugDescription();
+    void printBugToTerminal();
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::DOUBLEFREE;
+    }
+};
+
+class FileNeverCloseBug : public GenericBug{
+public:
+    FileNeverCloseBug(const EventStack &bugEventStack):
+          GenericBug(GenericBug::NEVERFREE, bugEventStack){  };
+
+    cJSON *getBugDescription();
+    void printBugToTerminal();
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::FILENEVERCLOSE;
+    }
+};
+
+class FilePartialCloseBug : public GenericBug{
+protected:
+    Set<std::string> conditionalFileClosePath;
+
+public:
+    FilePartialCloseBug(const EventStack &bugEventStack, Set<std::string> conditionalFileClosePath):
+          GenericBug(GenericBug::PARTIALLEAK, bugEventStack), conditionalFileClosePath(conditionalFileClosePath){ }
+
+    cJSON *getBugDescription();
+    void printBugToTerminal();
+
+    /// ClassOf
+    static inline bool classof(const GenericBug *bug)
+    {
+        return bug->getBugType() == GenericBug::FILEPARTIALCLOSE;
+    }
+};
+
+class SVFBugReport
+{
+public:
+    SVFBugReport() = default;
+    ~SVFBugReport();
+    typedef SVF::Set<GenericBug *> BugSet;
+
+protected:
+    BugSet bugSet;    // maintain bugs
+
+public:
+    /*
+     * function: pass bug type (i.e., GenericBug::BOA) and bug object as parameter,
+     *      it will add the bug into bugQueue.
+     * usage: addBug<GenericBug::BOA>(BufferOverflowBug(bugInst, 0, 10, 1, 11))
+     */
+    template <typename T>
+    void addBug(T bug)
```

Comment:
void addBug(eventstack, flag){

switch (xx)
case1:
xxxx

default:
assert(false && "no such kind of bug");

}